# Завдання
        
        Створити notebook homework_supervised_models.ipynb.

## Регуляризація на California Housing

### Завдання
1. Використовуйте датасет California Housing.
2. Візьміть ознаки MedInc, HouseAge, AveRooms, AveBedrms.
3. Побудуйте дві моделі: LinearRegression; Lasso(alpha=0.1).
Розділіть дані на train/test (80/20, random_state=42).
4. Для кожної моделі обчисліть R² на тестовій вибірці.
5. Порівняйте коефіцієнти моделей: скільки з них стали рівними 0 у Lasso.

In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Lasso, Ridge
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score

In [16]:
california_housing_df = pd.read_csv('https://raw.githubusercontent.com/dmitriy-shtefan/python-ai/main/data/california_housing.csv')
california_housing_df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847



* MedInc  -    Медіанний дохід
* HouseAge - 	Медіанний вік будинку
* AveRooms - 	Середня кількість кімнат
* AveBedrms - 	Середня кількість спалень
* Population -	Населення
* AveOccup - 	Середня кількість мешканців
* Latitude - 	Широта
* Longitude - 	Довгота
* MedHouseVal - 	Медіанна вартість житла

In [17]:
california_housing_df.describe()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


In [18]:
columns = ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms']
target = 'MedHouseVal'

X = california_housing_df[columns]
y = california_housing_df[target]  

print(X.shape, y.shape)

(20640, 4) (20640,)


In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

##### модель: LinearRegression

In [31]:
model_lr = LinearRegression()
model_lr.fit(X_train, y_train)
y_pred_lr = model_lr.predict(X_test)

mae_lr = mean_absolute_error(y_test, y_pred_lr)
mse_lr = mean_squared_error(y_test, y_pred_lr)
rmse_lr = np.sqrt(mse_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print(f"mae_lr = {mae_lr}"
      f"\nmse_lr = {mse_lr}"
      f"\nrmse_lr = {rmse_lr}"
      f"\nr2_lr = {r2_lr} - показує наскільки добре модель пояснює зміни y - target ")



mae_lr = 0.5804193614410437
mse_lr = 0.6435680467793014
rmse_lr = 0.8022269297270576
r2_lr = 0.5088800081000431 - показує наскільки добре модель пояснює зміни y - target 


##### модель: Lasso(alpha=0.1) 
лінійна регресія + правило 
- 0 - ознака не дуже важлива - допомагає спростити модель і уникнути переенавчання - overfitting
- alpha=0.1 - параметр штрафу, контролює на скільки сильно модель зменшує коефіцієнт. 0.1 - помірний штраф, де 1.0 - сильний штраф

In [32]:
model_lasso = Lasso(alpha=0.1)
model_lasso.fit(X_train, y_train)

y_pred_lasso = model_lasso.predict(X_test)

mae_lasso = mean_absolute_error(y_test, y_pred_lasso)
mse_lasso = mean_squared_error(y_test, y_pred_lasso)
rmse_lasso = np.sqrt(mse_lasso)
r2_lasso = r2_score(y_test, y_pred_lasso)

print(f"mae_lasso = {mse_lasso}\n"
      f"mse_lasso = {mse_lasso}\n"
      f"rmse_lasso = {rmse_lasso}\n"
      f"r2_lasso = {r2_lasso} - показує наскільки добре модель пояснює зміни y - target "
)

mae_lasso = 0.6640015192960609
mse_lasso = 0.6640015192960609
rmse_lasso = 0.8148628837393815
r2_lasso = 0.49328680562962846 - показує наскільки добре модель пояснює зміни y - target 


In [29]:
print(f"Модель Lasso - {model_lasso.coef_} - числа, які показують, як сильно кожна ознака вливає на передюачення ")

print()

print(f"Модель LinearRegression - {model_lr.coef_}")

Модель Lasso - [ 4.05075483e-01  1.63227560e-02 -4.03794138e-04  0.00000000e+00] - числа, які показують, як сильно кожна ознака вливає на передюачення 

Модель LinearRegression - [ 0.54526766  0.0160627  -0.2246198   1.11298793]


In [33]:
from tabulate import tabulate

comparison_data = {
    'Метрика': [
        'R2',
        'MAE',
        'MSE',
        'RMSE'
    ],
    'Linear Regression': [
        r2_lr,
        mae_lr,
        mse_lr,
        rmse_lr
    ],
    'Lasso': [
        r2_lasso,
        mae_lasso,
        mse_lasso,
        rmse_lasso
    ],
    'Різниця':
        [r2_lr - r2_lasso,
         mae_lr - mae_lasso,
         mse_lr - mse_lasso,
         rmse_lr - rmse_lasso]
}

print(tabulate(comparison_data, headers='keys', tablefmt='grid'))

+-----------+---------------------+----------+------------+
| Метрика   |   Linear Regression |    Lasso |    Різниця |
+===========+=====================+==========+============+
| R2        |            0.50888  | 0.493287 |  0.0155932 |
+-----------+---------------------+----------+------------+
| MAE       |            0.580419 | 0.610867 | -0.0304474 |
+-----------+---------------------+----------+------------+
| MSE       |            0.643568 | 0.664002 | -0.0204335 |
+-----------+---------------------+----------+------------+
| RMSE      |            0.802227 | 0.814863 | -0.012636  |
+-----------+---------------------+----------+------------+


по r2 різниця невелика, Lasso трохи гірше пояснює цільову змінну target, так як спрощує модель за рахунок штрафів.

Linear Regression - краще пояснює ціни на житло. 

In [ ]:
from tabulate import tabulate

zero_status = []
for c in model_lasso.coef_:
    if abs(c) < 1e-10:
        zero_status.append('Так')
    else:
        zero_status.append('Ні')
# abc - модуль числа 
# 1e-10 — це експоненційний запис числа: 1e-10 = 1 × 10⁻¹⁰ = 0.0000000001

coef_comperison = pd.DataFrame({
    'Ознка': columns,
    'LinearRegression': model_lr.coef_,
    'Lasso': model_lasso.coef_,
    'Різниця': model_lr.coef_ - model_lasso.coef_,

    'Lasso = 0?': zero_status


})

print(tabulate(coef_comperison, headers='keys', tablefmt='grid', showindex=False, floatfmt='.6f'))

+-----------+--------------------+-----------+-----------+--------------+
| Ознка     |   LinearRegression |     Lasso |   Різниця | Lasso = 0?   |
+===========+====================+===========+===========+==============+
| MedInc    |           0.545268 |  0.405075 |  0.140192 | Ні           |
+-----------+--------------------+-----------+-----------+--------------+
| HouseAge  |           0.016063 |  0.016323 | -0.000260 | Ні           |
+-----------+--------------------+-----------+-----------+--------------+
| AveRooms  |          -0.224620 | -0.000404 | -0.224216 | Ні           |
+-----------+--------------------+-----------+-----------+--------------+
| AveBedrms |           1.112988 |  0.000000 |  1.112988 | Так          |
+-----------+--------------------+-----------+-----------+--------------+


AveBedrms |           1.112988 |  0.000000 |  ознака видалена з моделі, вона не потрібна 

інші ознаки - чим більше показник 0,4 - середній дохід, тим вона важливіша.